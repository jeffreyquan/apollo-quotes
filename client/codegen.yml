schema: http://localhost:5000/graphql
generates:
  ./types.ts:
    plugins:
      - typescript
