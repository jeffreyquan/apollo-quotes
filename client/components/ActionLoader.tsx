import React from "react";
import ClipLoader from "react-spinners/ClipLoader";
import { ThemeConsumer } from "styled-components";

export const ActionLoader: React.FC = () => {
  return (
    <ThemeConsumer>
      {(theme) => <ClipLoader color={theme.actionloader} size={20} />}
    </ThemeConsumer>
  );
};
