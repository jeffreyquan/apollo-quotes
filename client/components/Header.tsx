import Link from "next/link";
import { useState, useEffect } from "react";
import styled from "styled-components";
import { IoIosRocket } from "react-icons/io";
import { Navbar } from "./Navbar";

const Logo = styled.div`
  font-size: 3rem;
  margin-right: auto;
  a {
    display: flex;
    align-items: center;
  }
`;

const StyledHeader = styled.header`
  position: sticky;
  background-color: ${(props) =>
    props.sticky ? props.theme.white : "transparent"};
  box-shadow: ${(props) =>
    props.sticky ? `0 0.2rem 0.5rem rgba(0,0,0,0.5)` : ""};
  top: 0;
  display: flex;
  align-items: center;
  transition: all 0.3s ease-in-out;
  line-height: ${(props) => (props.sticky ? "3rem" : "5rem")};
  max-height: 9rem;
`;

export const Header = () => {
  const [sticky, setSticky] = useState(false);

  useEffect(() => {
    window.addEventListener("scroll", () => {
      if (window.scrollY > 0) {
        setSticky(true);
      } else {
        setSticky(false);
      }
    });
  }, []);

  return (
    <StyledHeader sticky={sticky}>
      <Logo>
        <Link href="/">
          <a>
            <IoIosRocket />
            Apollo Quotes
          </a>
        </Link>
      </Logo>
      <Navbar />
    </StyledHeader>
  );
};
