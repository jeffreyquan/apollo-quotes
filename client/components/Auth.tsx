import { createContext, useState, useEffect } from "react";
import { useQuery, gql } from "@apollo/client";

const CURRENT_USER_QUERY = gql`
  query {
    userProfile {
      id
      name
      username
      email
      role
    }
  }
`;

export const AuthContext = createContext({
  checkingAuth: true,
  user: null,
  setUser: null,
});

export const AuthProvider = ({ children }) => {
  const [checkingAuth, setCheckingAuth] = useState(true);
  const [user, setUser] = useState(null);

  const { data, loading, error } = useQuery(CURRENT_USER_QUERY);

  useEffect(() => {
    if (typeof data !== undefined) {
      setCheckingAuth(false);
      if (data) {
        setUser(data.userProfile);
      }
    }
  }, [data]);

  const value = { checkingAuth, user, setUser };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};
