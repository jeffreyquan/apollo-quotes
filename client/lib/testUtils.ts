import { InMemoryCache } from "@apollo/client";

const NEW_QUOTE = "NEW_QUOTE";

function decodeCursor(encodedCursor: string) {
  return Buffer.from(encodedCursor, "base64").toString("ascii");
}

const testQuote = {
  id: "testquote123",
  author: "Forrest Gump, Forrest Gump",
  content:
    "Life was like a box of chocolates. You never know what you're gonna get.",
  image: "https://www.fillmurray.com/200/300",
  largeImage: "https://www.fillmurray.com/200/300",
  likes: [
    {
      id: "abc",
      user: { id: "abd123", username: "John", __typename: "User" },
      __typename: "Like",
    },
    {
      id: "abd",
      user: { id: "abe123", username: "Ben", __typename: "User" },
      __typename: "Like",
    },
  ],
  tags: [
    { id: "abe", name: "movie", __typename: "Tag" },
    { id: "abf", name: "life", __typename: "Tag" },
  ],
  submittedBy: {
    id: "abce1234",
    __typename: "User",
  },
  slug: "forrest-gump",
  __typename: "Quote",
};

const testCache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        quotes: {
          keyArgs: ["tag", "submittedBy", "likedBy"],
          merge: (
            existing = {
              __typename: "QuotesConnection",
              totalCount: 0,
              pageInfo: {
                endCursor: null,
                hasMore: true,
              },
              quotes: [],
            },
            incoming
          ) => {
            let existingCursor;
            let incomingCursor;

            if (
              existing.pageInfo.endCursor &&
              incoming.pageInfo.endCursor !== NEW_QUOTE
            ) {
              existingCursor = decodeCursor(existing.pageInfo.endCursor);

              incomingCursor = decodeCursor(incoming.pageInfo.endCursor);
            }

            if (incomingCursor > existingCursor) {
              return existing;
            }

            const newQuotes = incoming.quotes;

            let update;

            if (incoming.pageInfo.endCursor === NEW_QUOTE) {
              update = {
                pageInfo: existing.pageInfo,
                quotes: [...newQuotes, ...existing.quotes],
                totalCount: incoming.totalCount,
              };
            } else {
              update = {
                quotes: [...existing.quotes, ...newQuotes],
              };
            }

            return newQuotes.length > 0
              ? {
                  ...incoming,
                  ...update,
                }
              : existing;
          },
        },
      },
    },
    Quote: {
      fields: {
        likes: {
          merge(existing = [], incoming) {
            return incoming;
          },
        },
      },
    },
  },
});

const testNewQuote = {
  id: "abcd123",
  author: "CS",
  content: "Hello World",
  image: "",
  largeImage: "",
  tags: [
    { id: "tag123", name: "power" },
    { id: "tag124", name: "life" },
  ],
  likes: [],
  slug: "hello-world-cs",
};

export { testCache, testQuote, testNewQuote };
