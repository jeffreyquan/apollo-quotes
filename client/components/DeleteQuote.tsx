import React from "react";
import { gql, useMutation } from "@apollo/client";
import { BsExclamationCircle } from "react-icons/bs";
import { useRouter } from "next/router";
import { ActionLoader } from "./ActionLoader";

const DELETE_QUOTE_MUTATION = gql`
  mutation DELETE_QUOTE_MUTATION($id: ID!) {
    deleteQuote(id: $id) {
      id
    }
  }
`;

interface DeleteQuoteProps {
  id: string;
}

export const DeleteQuote: React.FC<DeleteQuoteProps> = ({ id, children }) => {
  const [deleteQuote, { error, loading }] = useMutation(DELETE_QUOTE_MUTATION, {
    variables: { id },
    update(cache, { data: { deleteQuote } }) {
      cache.modify({
        fields: {
          quotes(
            existing = {
              __typename: "QuotesConnection",
              totalCount: 0,
              pageInfo: {
                endCursor: null,
                hasMore: true,
              },
              quotes: [],
            },
            { readField }
          ) {
            const updatedQuotes = existing.quotes.filter(
              (quoteRef) => readField("id", quoteRef) !== deleteQuote.id
            );
            return {
              ...existing,
              quotes: [...updatedQuotes],
            };
          },
        },
      });
    },
  });

  if (loading) return <ActionLoader />;
  if (error) return <BsExclamationCircle />;

  const router = useRouter();

  const handleClick = async (e) => {
    e.preventDefault();
    try {
      await deleteQuote();
      router.push({
        pathname: "/quotes",
        query: { delete: "success" },
      });
    } catch (err) {
      return <BsExclamationCircle />;
    }
  };

  return (
    <div
      data-testid="deleteButton"
      onClick={(e) => {
        if (confirm("Please confirm you want to delete this quote.")) {
          handleClick(e);
        }
      }}
    >
      {children}
    </div>
  );
};
