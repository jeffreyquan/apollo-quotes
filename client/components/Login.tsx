import { useState, useEffect, useContext } from "react";
import { useRouter } from "next/router";
import { gql, useMutation } from "@apollo/client";
import { AuthContext } from "./Auth";
import { useForm } from "../lib/useForm";
import { Form } from "../styles/Form";
import { FormContainer } from "../styles/FormContainer";
import { FormTitle } from "../styles/FormTitle";

const LOGIN_MUTATION = gql`
  mutation LOGIN_MUTATION($email: String!, $password: String!) {
    login(email: $email, password: $password) {
      id
      name
      email
      username
    }
  }
`;

export const Login = () => {
  const { inputs, handleChange, resetForm } = useForm({
    email: "",
    password: "",
  });

  const { email, password } = inputs;

  const [loadingPage, setLoadingPage] = useState(true);

  let { user, setUser } = useContext(AuthContext);

  const router = useRouter();

  useEffect(() => {
    if (user) {
      if (router.query.redirect) {
        router.back();
      } else {
        router.push("/");
      }
    } else {
      setLoadingPage(false);
    }
  }, [user]);

  const [login, { error, loading }] = useMutation(LOGIN_MUTATION, {
    variables: inputs,
  });

  // TODO: create error component to display any login errors
  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const res = await login();
      setUser(res.data.login);
    } catch (err) {
      console.log(err);
    }
  };

  return !loadingPage ? (
    <FormContainer>
      <Form onSubmit={handleSubmit}>
        <FormTitle>Login</FormTitle>
        <fieldset disabled={loading} aria-busy={loading}>
          <label htmlFor="email">
            Email
            <input
              type="email"
              name="email"
              placeholder="Email"
              value={email}
              onChange={handleChange}
              autoComplete="email"
            />
          </label>
          <label htmlFor="password">
            Password
            <input
              type="password"
              name="password"
              placeholder="Password"
              value={password}
              onChange={handleChange}
              autoComplete="current-password"
            />
          </label>
          <input type="submit" value="Sign In" />
        </fieldset>
      </Form>
    </FormContainer>
  ) : (
    <div>Loading....</div>
  );
};
