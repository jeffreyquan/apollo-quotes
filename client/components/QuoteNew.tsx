import React from "react";
import { gql, useMutation } from "@apollo/client";
import { useEffect, useRef, useState } from "react";
import { useRouter } from "next/router";
import { MdAddCircle } from "react-icons/md";
import { useForm } from "../lib/useForm";
import { Form } from "../styles/Form";
import { FormTitle } from "../styles/FormTitle";
import { FormContainer } from "../styles/FormContainer";
import { QuoteTag } from "../styles/QuoteTag";
import { AddIconStyles } from "../styles/AddIconStyles";
import { Message } from "./Message";

export const CREATE_QUOTE_MUTATION = gql`
  mutation CREATE_QUOTE_MUTATION(
    $content: String!
    $author: String!
    $image: Upload
    $tags: [String]
  ) {
    createQuote(
      content: $content
      author: $author
      image: $image
      tags: $tags
    ) {
      id
      author
      content
      image
      largeImage
      tags {
        id
        name
      }
      likes {
        id
        user {
          id
          username
        }
      }
      slug
    }
  }
`;

export const QuoteNew: React.FC = () => {
  const { inputs, handleChange, updateInputs } = useForm({
    content: "",
    author: "",
    image: "",
    tags: [],
  });

  const [tagInput, setTagInput] = useState("");

  const [errorMessage, setErrorMessage] = useState("");

  const { content, author, image, tags } = inputs;

  const router = useRouter();

  // TODO: refetch all quotes after mutation
  const [createQuote, { error, loading }] = useMutation(CREATE_QUOTE_MUTATION, {
    variables: inputs,
    update(cache, { data: { createQuote } }) {
      cache.modify({
        fields: {
          quotes(
            existing = {
              __typename: "QuotesConnection",
              totalCount: 0,
              pageInfo: {
                endCursor: null,
                hasMore: true,
              },
              quotes: [],
            },
            { readField }
          ) {
            const newQuoteRef = cache.writeFragment({
              data: createQuote,
              fragment: gql`
                fragment NewQuote on Quote {
                  id
                  author
                  content
                  image
                  largeImage
                  tags {
                    id
                    name
                  }
                  likes {
                    id
                    user {
                      id
                      username
                    }
                  }
                  slug
                }
              `,
            });

            if (
              existing.quotes.some(
                (ref) => readField("id", ref) === createQuote.id
              )
            ) {
              return existing;
            }

            return {
              ...existing,
              quotes: [newQuoteRef, ...existing.quotes],
            };
          },
        },
      });
    },
  });

  if (error)
    setErrorMessage("Quote was unsuccessfully created. Please try again.");

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const res = await createQuote();
      const slug = res.data.createQuote.slug;
      router.push("/quotes/[slug]", `/quotes/${slug}`);
    } catch (err) {
      setErrorMessage(err.message);
    }
  };

  const timeoutId = useRef<number>();

  useEffect(() => {
    if (errorMessage) {
      timeoutId.current = window.setTimeout(function () {
        setErrorMessage("");
      }, 3000);
    }
  }, [errorMessage]);

  const addTag = () => {
    const tagName = tagInput.toLowerCase();
    if (!tags.includes(tagName)) {
      updateInputs({
        tags: [...tags, tagName],
      });
      setTagInput("");
    }
  };

  const removeTag = (index) => {
    const updatedTags = [...tags];
    updatedTags.splice(index, 1);
    updateInputs({
      tags: updatedTags,
    });
  };

  return (
    <FormContainer>
      <Message error={errorMessage ? true : false}>{errorMessage}</Message>
      <Form onSubmit={handleSubmit}>
        <FormTitle>Submit a new quote</FormTitle>
        <fieldset disabled={loading} aria-busy={loading}>
          <label htmlFor="content">Content</label>
          <textarea
            id="content"
            name="content"
            placeholder="When nothing goes right, go left."
            rows={4}
            value={content}
            onChange={handleChange}
          />
          <label htmlFor="author">Author</label>
          <input
            id="author"
            type="text"
            name="author"
            placeholder="JR Smith"
            value={author}
            onChange={handleChange}
          />
          <label htmlFor="file">Image</label>
          <input
            type="file"
            id="file"
            name="image"
            placeholder="Upload an image"
            onChange={handleChange}
            value={image}
            required
          />
          <label htmlFor="tag">Tags</label>
          <div className="input__group">
            <input
              id="tag"
              name="tag"
              type="text"
              value={tagInput}
              onChange={(e) => setTagInput(e.target.value)}
              placeholder="power"
            />
            <AddIconStyles data-testid="addTag" onClick={() => addTag()}>
              <MdAddCircle />
            </AddIconStyles>
          </div>
          <div>
            {tags.length > 0 &&
              tags.map((tag, i) => {
                return (
                  <QuoteTag
                    key={`${tag}-${i}`}
                    className="edit"
                    onClick={() => removeTag(i)}
                  >
                    {tag}
                  </QuoteTag>
                );
              })}
          </div>
          <input type="submit" value="Submit" />
        </fieldset>
      </Form>
    </FormContainer>
  );
};
