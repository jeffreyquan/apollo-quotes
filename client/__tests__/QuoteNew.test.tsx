import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockedProvider } from "@apollo/client/testing";
import { useRouter } from "next/router";
import { CREATE_QUOTE_MUTATION, QuoteNew } from "../components/QuoteNew";
import { testNewQuote } from "../lib/testUtils";

const mockRouterPush = jest.fn();

jest.mock("next/router", () => ({
  ...jest.requireActual("next/router"),
  useRouter: () => ({
    push: mockRouterPush,
  }),
}));

const mocks = [
  {
    request: {
      query: CREATE_QUOTE_MUTATION,
      variables: {
        content: "Hello World",
        author: "CS",
        image: "",
        tags: ["power", "life"],
      },
    },
    result: {
      data: {
        createQuote: {
          __typename: "Quote",
          id: "abcd123",
          author: "CS",
          content: "Hello World",
          image: "",
          largeImage: "",
          tags: [
            { id: "tag123", name: "power", __typename: "Tag" },
            { id: "tag124", name: "life", __typename: "Tag" },
          ],
          likes: [],
          slug: "hello-world-cs",
        },
      },
    },
  },
];

describe("QuoteNew", () => {
  it("renders all inputs and labels", () => {
    const { getByRole, getByLabelText } = render(
      <MockedProvider>
        <QuoteNew />
      </MockedProvider>
    );

    expect(
      getByRole("heading", { name: "Submit a new quote" })
    ).toBeInTheDocument();

    expect(getByLabelText("Content")).toBeInTheDocument();

    expect(getByLabelText("Author")).toBeInTheDocument();

    expect(getByLabelText("Image")).toBeInTheDocument();

    expect(getByLabelText("Tags")).toBeInTheDocument();

    expect(getByRole("button", { name: "Submit" })).toBeInTheDocument();

    expect(
      screen.getByPlaceholderText("When nothing goes right, go left.")
    ).toBeInTheDocument();

    expect(screen.getByPlaceholderText("JR Smith")).toBeInTheDocument();

    expect(screen.getByPlaceholderText("Upload an image")).toBeInTheDocument();
  });

  it("creates a new quote", async () => {
    const { getByPlaceholderText, getByRole } = render(
      <MockedProvider mocks={mocks}>
        <QuoteNew />
      </MockedProvider>
    );

    const router = useRouter();

    await userEvent.type(
      getByPlaceholderText("When nothing goes right, go left."),
      testNewQuote.content
    );

    await userEvent.type(getByPlaceholderText("JR Smith"), testNewQuote.author);

    await userEvent.type(
      getByPlaceholderText("power"),
      testNewQuote.tags[0].name
    );

    await userEvent.click(screen.getByTestId("addTag"));

    await userEvent.type(
      getByPlaceholderText("power"),
      testNewQuote.tags[1].name
    );

    await userEvent.click(screen.getByTestId("addTag"));

    await userEvent.type(getByPlaceholderText("power"), "chicken");

    await userEvent.click(screen.getByTestId("addTag"));

    expect(screen.getByText("chicken")).toBeInTheDocument();

    await userEvent.click(screen.getByText("chicken"));

    expect(screen.queryByText("chicken")).not.toBeInTheDocument();

    await userEvent.click(getByRole("button", { name: "Submit" }));

    await waitFor(() =>
      expect(router.push).toHaveBeenCalledWith(
        "/quotes/[slug]",
        `/quotes/${testNewQuote.slug}`
      )
    );
  });
});
