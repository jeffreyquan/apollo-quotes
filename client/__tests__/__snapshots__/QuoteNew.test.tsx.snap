// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteNew renders all inputs and labels 1`] = `
<div>
  <div
    class="FormContainer-sc-1xw7mz9-0 gTcmkT"
  >
    <div
      class="Message__MessageStyles-sc-1wz8v68-0 kcegBB"
    />
    <form
      class="Form-sc-1stku61-0 jaHlPk"
    >
      <h3
        class="FormTitle-i1lio3-0 bihtdU"
      >
        Submit a new quote
      </h3>
      <fieldset
        aria-busy="false"
      >
        <label
          for="content"
        >
          Content
        </label>
        <textarea
          id="content"
          name="content"
          placeholder="When nothing goes right, go left."
          rows="4"
        />
        <label
          for="author"
        >
          Author
        </label>
        <input
          id="author"
          name="author"
          placeholder="JR Smith"
          type="text"
          value=""
        />
        <label
          for="file"
        >
          Image
        </label>
        <input
          id="file"
          name="image"
          placeholder="Upload an image"
          required=""
          type="file"
        />
        <label
          for="tag"
        >
          Tags
        </label>
        <div
          class="input__group"
        >
          <input
            id="tag"
            name="tag"
            placeholder="power"
            type="text"
            value=""
          />
          <div
            class="AddIconStyles-sc-3fm2p2-0 buLSIj"
            data-testid="addTag"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
              />
            </svg>
          </div>
        </div>
        <div />
        <input
          type="submit"
          value="Submit"
        />
      </fieldset>
    </form>
  </div>
</div>
`;
